aten/src/ATen/native/Blas.cpp

```diff
@@ -19,86 +36,72 @@ constexpr inline bool lda_cond(int64_t m, int64_t n, int64_t lda) {
-              beta, self.scalar_type(), c10::nullopt /* layout */, at::kCPU, c10::nullopt /* pin_memory */));
+              beta_, self.scalar_type(), c10::nullopt /* layout */, at::kCPU, c10::nullopt /* pin_memory */));
     }
   } else {
-    if (!result.is_same(self_)) {
-      at::native::copy_(result, self_);
+    if (!result.is_same(self_) && betaval != 0.0) { //if beta is 0, result contents is ignored
+      at::native::copy_(const_cast<Tensor&>(result), self_);
     }
     if (result.numel() != 0) {
-      at::_addmv_impl_(result, self_, mat, vec, beta, alpha);
+      auto r_stride = result.stride(0);
+      AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16, mat.scalar_type(), "addmv_impl_cpu", [&] {
+        auto beta = beta_.to<scalar_t>();
+        auto alpha = alpha_.to<scalar_t>();
+        if (mat.stride(0) == 1 && lda_cond(mat.size(0), mat.size(1), mat.stride(1))) {
+          gemv<scalar_t>('n', mat.size(0), mat.size(1), alpha, mat.data_ptr<scalar_t>(), mat.stride(1),
+              vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
+        }
+        else if (mat.stride(1) == 1 && lda_cond(mat.size(1), mat.size(0), mat.stride(0))) {
+          gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, mat.data_ptr<scalar_t>(), mat.stride(0),
+              vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
+        }
+        else {
+          Tensor cmat = mat.contiguous();
+          gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, cmat.data_ptr<scalar_t>(), cmat.stride(0),
+              vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
+        }
+      });
     }
   }
-
-  } // scope of NoNamesGuard
-  at::namedinference::propagate_names_for_addmv(result, mat, vec, self);
-  return result;
-}
-
-Tensor addmv(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha) {
-  Tensor result = at::empty({mat.size(0)}, mat.options());
-  return native::addmv_out(self, mat, vec, beta, alpha, result);
-}
-
-Tensor &addmv_(Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha) {
-  return native::addmv_out(self, mat, vec, beta, alpha, self);
 }
 
 Tensor &mv_out(const Tensor &self, const Tensor &vec, Tensor& result) {
-  return native::addmv_out(result, self, vec, 0, 1, result);
+  //self arg sent to addmv_out cannot be resized
+  //here we use result as self argument for addmv, and result is user supplied and can be wrong size
+  //it's not a hard error, because we allow resizing result, but it becomes a hard error
+  //in addmv, because addmv expects self to satisfy proper conditions
+  //to avoid this, supply correctly sized self, its contents doesn't matter because beta is 0
+  if (result.dim() > 1 || (result.numel() != self.size(0) || result.numel() !=1)) {
+    Tensor self_addmv = at::empty({self.size(0)}, self.options());
```

Comment:
Ick, we probably shouldn't implement it this way (preexisting condition, carry on...)